### ISSUE TITLE
AttributeError: module 'lib' has no attribute 'load_config'

### ISSUE THREAD

**Issue by imasojeong** [Yura52/tabular-dl-revisiting-models#22]
I have already complete the "PyTorch environment" section and complete the "Data" section.

And I have completed the following.
After Step 2, but before Step 3, make sure that:
you are at the repository root
you have activated the corresponding conda environment
you did export CUDA_VISIBLE_DEVICES=0

But it's still the same error.

**Comment by imasojeong** [comment]
I did set PYTHONPATH=$PROJECT_DIR and did python bin/mlp.py draft/check_environment.toml and it was solved!

**Comment by judy19980326** [comment]
I encountered the same problem, when I executed python bin/mlp.py draft/check_environment. Toml times this error, may I ask if this is the problem of $PROJECT_DIR environment setting, I am trying to reproduce this experiment using Windows system, There are some conflicts in the environment, and I don't know how to solve them. Can you help me @imasojeong @Yura52

**Comment by Yura52** [comment]
Hi @judy19980326 , are you following the instructions given in [this section](https://github.com/Yura52/tabular-dl-revisiting-models#3-setup-the-environment)? Creating a separate virtual environment is especially important to avoid conflicts. Is `PROJECT_DIR` pointing to the root of the repository? Also, you can try [setting up WSL](https://learn.microsoft.com/en-us/windows/wsl/install) to avoid windows-specific problems (unfortunately, I am not familiar with setting up python environments on Windows).

**Comment by judy19980326** [comment]
> 

@Yura52，Thank you very much for your reply. I got the inspiration to carefully study and modify some adjustments of the Windows version about environment variables. I am very glad that this problem has been solved, but a new problem has come up and it works successfully: 

(revisiting-models) C:\Users\zhuyuedi\dataset>python bin/mlp.py draft/check_environment.toml
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = =
Config: C:\Users\zhuyuedi\dataset\draft\check_environment.toml
Output: C:\Users\zhuyuedi\dataset\draft\check_environment
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 
Already done!

A folder is also generated, but the contents of the folder are empty

**Comment by Yura52** [comment]
Great! Could you please share the windows-specific changes that you had to perform to make everything work?

Regarding the empty folder, the following happens:
- when you run the script with the config, an empty folder is created next to the config to store the results
- this folder remains even if the script failed
- next time you run the script with the same config, it detects the existing folder and immidiately exits to avoid accidental rewriting. If you are sure that you want to rewrite the existing folder, use ` --force` flag: `python bin/some_script.py some/config.toml --force`. If you want the existing folder to stay untouched, then copy the config to a file with a different name and use the new copy.

**Comment by judy19980326** [comment]
> 
     Thank you very much for your help. I finally compromised and installed dual system for my computer. Later, I also encountered the same problem. The reason is probably the compatibility problem between cuda and pytorch version of my graphics card RTX3060.I'm glad that it was resolved after upgrading the pytorch version.After reading your two papers 《Revisiting Deep Learning Models for Tabular Data》《On Embeddings for Numerical Features in Tabular Deep Learning》and reproducing your experiments, I really benefited a lot, and also strengthened my research direction.
When applying my own dataset, I encountered some problems:
    My dataset contains cat Features and num Features, similar to the Adult dataset in your paper. If I want to apply ft_ Transformer model is used to train it. My files are in CSV format. How should I process them as npy files. I searched through the folders, but unfortunately I didn't find any code related to data preprocessing. I would appreciate your help.

**Comment by Yura52** [comment]
@judy19980326 Does [this comment](https://github.com/Yura52/tabular-dl-revisiting-models/issues/3#issuecomment-1068124334) answer your questions? Regarding the specific tools you can use, you can read your data with pandas, and save it in the required format using [`numpy.save`](https://numpy.org/doc/stable/reference/generated/numpy.save.html)

**Comment by judy19980326** [comment]
> 

Thank you Yura, it is very helpful to me. After continuous debugging, it will be successful soon. Now I have a problem that the data type of my N_train,N_val and N_test is float64. Execution times error:

RuntimeError: expected scalar type Float but found Double

I know that the default data type of torch is float32. Will it affect if I reduce the precision of float64 to float32? Please help me how to deal with it.Thank you again for your help. @Yura52

**Comment by Yura52** [comment]
@judy19980326 In most cases, working with f32 instead of f64 is fine and gives similar results. Anyway, I would give it a try and see if the results are good enough.

**Comment by judy19980326** [comment]
> @judy19980326 Does [this comment](https://github.com/Yura52/tabular-dl-revisiting-models/issues/3#issuecomment-1068124334) answer your question? Regarding the specific tools you can use, you can read your data with pandas, and save it in the required format using [`numpy.save`](https://numpy.org/doc/stable/reference/generated/numpy.save.html)

Just after I converted them to float32, I encountered a new setback. My dataset contains cat, num features, and multiple classification problems. On my next execution:

.. Aten/SRC/aten/native/cuda/Indexing. The cu: 975: indexSelectLargeIndex: block: 0, [16], thread: [63,0,0] Assertion 'srcIndex < srcSelectDimSize' failed.

Traceback (most recent call last):

File "bin/ft_transformer.py", line 508, in

metrics, predictions = evaluate([lib.VAL, lib.TEST])

File "/ home/zhuyuedi anaconda3 / envs/revisiting - models/lib/python3.8 / site - packages/torch/autograd/grad_mode py", line 27, in decorate_context

return func(*args, **kwargs)

File "bin/ft_transformer.py", line 432, in evaluate

[

File "bin/ft_transformer.py", line 433, in

apply_model(part, idx)

File "bin/ft_transformer.py", line 416, in apply_model

return model(

File "/ home/zhuyuedi anaconda3 / envs/revisiting - models/lib/python3.8 / site - packages/torch/nn/modules/module. The p y", line 1130, in _call_impl

return forward_call(*input, **kwargs)

File "bin/ft_transformer.py", line 276, in forward

x_residual = layer['attention'](

File "/ home/zhuyuedi anaconda3 / envs/revisiting - models/lib/python3.8 / site - packages/torch/nn/modules/module. The p y", line 1130, in _call_impl

return forward_call(*input, **kwargs)

File "bin/ft_transformer.py", line 140, in forward

x = attention @ self._reshape(v)

RuntimeError: CUDA error: device-side assert triggered

CUDA kernel errors might be asynchronously reported at some other API call,so the stacktrace below might be incorrect.

For debugging consider passing CUDA_LAUNCH_BLOCKING=1.





Everything in my info file is correct (including

"task_type": "multiclass",

"n_num_features": 39,

"n_cat_features": 3,

"train_size": 224319,

"val_size": 56080,

"test_size": 51535,

"n_classes": 10

})



The interesting thing is that when I change n_cat_features to 0, and I take away C_train, C_val,C_test, the model runs smoothly. I don't know whether to be grateful or continue to solve the problem with cat_ Error in features

**Comment by Yura52** [comment]
> Just after I converted them to float32, I encountered a new setback

Things like this can happen when you have categorical features with categories presented in Val or test sets, but not in the train set. Can you check that?

**Comment by judy19980326** [comment]
> Things like this can happen when you have categorical features with categories presented in Val or test sets, but not in the train set. Can you check that?

Thank you very much for your prompt. Due to the time difference, I failed to reply in time. I will try to adjust it immediately.

**Comment by judy19980326** [comment]
> > Just after I converted them to float32, I encountered a new setback
> 
> Things like this can happen when you have categorical features with categories presented in Val or test sets, but not in the train set. Can you check that?

I finally found a controversial point. When I look at your only adult dataset with cat_features, I find that C_trainval.npy, N_trainval.npy, and y_trainval.npy are missing, but I can't find that they are involved in the operation through the code. Do these three npy files have the meaning of existence

**Comment by Yura52** [comment]
No, those files are not used.

**Comment by judy19980326** [comment]
> No, those files are not used.

OK，I'm more likely to blame the evaluate () phase for not handling the evaluation function with cat_features. But I'm a bit of a code fool and can't figure that out

**Comment by Yura52** [comment]
In [this message](https://github.com/Yura52/tabular-dl-revisiting-models/issues/22#issuecomment-1287815624), I mean running the following script:

```python
import numpy as np

C_train = np.load('data/<YOUR DATASET>/C_train.npy')
C_val = np.load('data/<YOUR DATASET>/C_val.npy')
C_test = np.load('data/<YOUR DATASET>/C_test.npy')

n_cat_features = C_train.shape[1]
ok = True
for i in range(n_cat_features):
    train_categories = set(C_train[:, i])
    val_categories = set(C_val[:, i])
    test_categories = set(C_test[:, i])
    if not (val_categories <= train_categories):
        ok = False
        print(f'ERROR: {i} val')
    if not (test_categories <= train_categories):
        ok = False
        print(f'ERROR: {i} test')
if ok:
    print('OK')
```

What does this script print for your dataset?

**Comment by judy19980326** [comment]
> 
Thank you very much, the error is indeed here, I run the above code error:

ERROR: 2 val

ERROR: 2 test

May I ask what might be the cause of this? I am sorry that I feel very confused.

The only data problems I can think of are:

['unas' '-' 'INT']

['tcp' '-' 'CON']

['tcp' '-' 'FIN']

.

['tcp' 'http' 'FIN']

['unas' '-' 'INT']

['tcp' '-' 'FIN']]

(224319, 3)


(224319, 3)


[['udp' '-' 'INT']

['udp' '-' 'INT']

['tcp' '-' 'FIN']

.

['tcp' '-' 'FIN']

['udp' '-' 'INT']

['tcp' 'ssh' 'FIN']]

(56080, 3)


(56080, 3)


[['tcp' '-' 'FIN']

['unas' '-' 'INT']

['tcp' '-' 'FIN']

.

['udp' 'dns' 'INT']

['udp' 'dns' 'INT']

['tcp' 'http' 'FIN']]

(51535, 3)


(51535, 3)


cat_features There are some unknown features annotated with '-', but it also exists in C_train, C_val, C_test

**Comment by judy19980326** [comment]
> ```python
> import numpy as np
> 
> C_train = np.load('data/<YOUR DATASET>/C_train.npy')
> C_val = np.load('data/<YOUR DATASET>/C_val.npy')
> C_test = np.load('data/<YOUR DATASET>/C_test.npy')
> 
> n_cat_features = C_train.shape[1]
> ok = True
> for i in range(n_cat_features):
>     train_categories = set(C_train[:, i])
>     val_categories = set(C_val[:, i])
>     test_categories = set(C_test[:, i])
>     if not (val_categories <= train_categories):
>         ok = False
>         print(f'ERROR: {i} val')
>     if not (test_categories <= train_categories):
>         ok = False
>         print(f'ERROR: {i} test')
> if ok:
>     print('OK')
> ```

I seem to have found the cause of the problem. I print the feature category through the code you provided:

3

I equal to zero

{

I equal to 2

{'REQ', 'CON', 'PAR', 'RST', 'ECO', 'INT', 'FIN', 'no'}

= = = = = = = = = = = = = = = = = =

{'REQ', 'ACC', 'CLO', 'CON', 'RST', 'ECO', 'INT', 'FIN'}

= = = = = = = = = = = = = = = = = =

{'URN', 'REQ', 'ACC', 'CON', 'ECO', 'RST', 'INT', 'FIN'}

ERROR: 2 val

ERROR: 2 test



What can I do if I find a feature that exists in C_val and C_test but does not exist in C_Train? Can I be clever and manually modify the data that does not exist in C_Train into some data......

**Comment by Yura52** [comment]
(please, create a separate issue if there will be further questions)

There is no "easy" way to solve the problem, I described several possible strategies [here](https://github.com/Yura52/rtdl/issues/37#issuecomment-1237433858).

**Comment by judy19980326** [comment]
> 

Well, thank you very much indeed
